libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -167,6 +168,255 @@ _rtld_relocate_nonplt_self(Elf_Dyn *dynp, Elf_Auxinfo *aux)
+
+	ptraddr_t key = (ptraddr_t)dst;
+
+	if (cur->key) {
+
+		ptraddr_t lim;
+		asm ("gclim	%0, %1" : "=r" (lim) : "C" (table));
+
+		size_t old_len = (flags + 1) << DEFAULT_FLAG_WIDTH;
+
+		flags |= flags + 1;
+		size_t new_len = (flags + 1) << DEFAULT_FLAG_WIDTH;
+		tramp_stk_table_t new_t = xcalloc(new_len, sizeof(*new_t));
+
+		asm ("scflgs	%0, %1, %2" : "=C" (new_t) : "C" (new_t), "r" (flags << 56));
+
+		for (size_t i = 0; i < old_len; ++i) {
+			if (!table[i].key)
+				continue;
+
+			size_t offset = HASH_KEY(key, flags);
+
+			// Guaranteed to terminate
+			while (new_t[offset].key) {
+				offset = (offset - 1) & (new_len - 1);
+			}
+
+			new_t[offset].key = table[i].key;
+			new_t[offset].stk = table[i].stk;
+		}
+
+		asm ("msr	ctpidr_el0, %0" :: "C" (new_t));
+		free(table);
+
+	} else {
+
+		size_t size = 0x40000 * getpagesize();
+		char *stk = mmap(NULL,
+				 size,
+				 PROT_READ | PROT_WRITE,
+				 MAP_ANON | MAP_PRIVATE | MAP_STACK,
+				 -1, 0);
+		if (stk == MAP_FAILED)
+			rtld_die();
+		stk = cheri_clearperm(stk, CHERI_PERM_EXECUTIVE) + size;
+		struct {
+			uint8_t generation;
+			uintptr_t top;
+		} *metadata = (void *)stk;
+		metadata[-1].generation = 0;
+		metadata[-1].top = (uintptr_t)&metadata[-1];
+
+		cur->key = key;
+		cur->stk = stk;
+
+		struct Struct_Stack_Entry *entry = xmalloc(sizeof(*entry));
+		entry->stack = stk;
+
+		lockinfo.wlock_acquire(dst->stackslock);
+		SLIST_INSERT_HEAD(&dst->stacks, entry, link);
+		lockinfo.lock_release(dst->stackslock);
+	}
+}
+
+#else /* ! HASHTABLE_STACK_SWITCHING */
+
+#define DEFAULT_STACK_TABLE_SIZE 2
+
+typedef uintptr_t *tramp_stk_table_t;
+
+static void
+get_rstk(uint32_t index, tramp_stk_table_t table, void *target)
+{
+	size_t len = cheri_getlen(table) / sizeof(*table);
+
+	if (index < len) {
+
+		size_t size = 0x40000 * getpagesize();
+		char *stk = mmap(NULL,
+				 size,
+				 PROT_READ | PROT_WRITE,
+				 MAP_ANON | MAP_PRIVATE | MAP_STACK,
+				 -1, 0);
+		if (stk == MAP_FAILED)
+			rtld_die();
+		stk = cheri_clearperm(stk, CHERI_PERM_EXECUTIVE) + size;
+		struct {
+			uint8_t generation;
+			uintptr_t top;
+		} *metadata = (void *)stk;
+		metadata[-1].generation = 0;
+		metadata[-1].top = (uintptr_t)&metadata[-1];
+
+		table[index] = (uintptr_t)stk;
+
+		struct Struct_Stack_Entry *entry = xmalloc(sizeof(*entry));
+		entry->stack = stk;
+
+		Obj_Entry *dst = obj_from_addr(target);
+		lockinfo.wlock_acquire(dst->stackslock);
+		SLIST_INSERT_HEAD(&dst->stacks, entry, link);
+		lockinfo.lock_release(dst->stackslock);
+
+	} else {
+
+		size_t new_len = len * 2;
+		tramp_stk_table_t new_t = xcalloc(new_len, sizeof(*new_t));
+		if (!new_t)
+			rtld_die();
+		assert(cheri_getlen(new_t) / sizeof(*new_t) == new_len);
+
+		for (size_t i = 0; i < len; ++i)
+			new_t[i] = table[i];
+
+		asm ("msr	ctpidr_el0, %0" :: "C" (new_t));
+		free(table);
+	}
+}
+
+#endif
+
+static void (*_thr_thread_entry)(struct pthread *);
```

Comment:
I think this should be OK, because this variable is written to only once in `_rtld_thread_start_tramp`, when the first thread is being created. So there's presumable no threading issue here?

Reply:
Right, I see, and you do it lazily for performance reasons? Seems a little unnecessary but I guess it's fine.

Reply:
`_rtld_thread_start_tramp` is called by `libthr` to inject its `thread_start` to `rtld`. I don't think `rtld` can directly link to `thread_start` in another library (since it only does relative relocations during self-relocation at startup)?

Reply:
It _can_ (see _pli_rtli_version), though this probably belongs in thr_init.c along with all the other dynamic reconfiguration of rtld for threading